File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.py
from typing import List

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        n = len(nums) - 1
        ans = 0

        for shift in range(32):
            count = 0

            # Count set bits in array
            for num in nums:
                count += (num & (1 << shift))

            # Subtract expected set bits from 1..n
            for i in range(1, n + 1):
                count -= (i & (1 << shift))

            # If difference remains, it must come from the duplicate number
            if count > 0:
                ans |= (1 << shift)

        return ans

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int n = nums.length - 1;
        int ans = 0;

        for (int shift = 0; shift < 32; shift++) {
            int count = 0;

            // Count set bits in array
            for (int num : nums) {
                count += (num & (1 << shift));
            }

            // Count expected set bits
            for (int i = 1; i <= n; i++) {
                count -= (i & (1 << shift));
            }

            // If difference remains, must be from duplicate number
            if (count > 0)
                ans |= (1 << shift);
        }
        return ans;
    }
}